feat: Territory autocomplete

m0wer opened the pull request:
Closes #992.

## Description

Added territory autocomplete functionality to enhance the user experience when entering territory names. This feature provides suggestions as users type territory names using the tilde (~) prefix, similar to how the existing user mention system works with @ mentions.

Key changes:
- Added a new GraphQL query `subSuggestions` to fetch territory name suggestions based on similarity to input
- Created a `TerritorySuggest` component that mirrors the existing `UserSuggest` component
- Integrated territory suggestions in the markdown input field
- Added appropriate fragment for territory suggestions

## Screenshots

![image](https://github.com/user-attachments/assets/5b06622e-80c5-411e-ab39-e05f92cdb6d8)

![image](https://github.com/user-attachments/assets/182388a8-4920-4b2a-87ec-34b2e4616b99)

## Additional Context

The implementation follows the same pattern as the existing user mention system. Territory suggestions appear when typing "~" followed by any text, and users can navigate/select suggestions using arrow keys, Enter, or mouse click.

## Checklist

**Are your changes backwards compatible? Please answer below:**
Yes, this feature adds new functionality without breaking existing code. The existing text input behavior remains unchanged.

**On a scale of 1-10 how well and how have you QA'd this change and any features it might affect? Please answer below:**
8/10. I've tested the territory autocomplete functionality with various input scenarios including empty search, partial matches, and full territory names. I've also verified that it doesn't interfere with the existing user mention functionality.

**For frontend changes: Tested on mobile, light and dark mode? Please answer below:**
Yes, tested on mobile devices and in both light and dark mode. The dropdown positioning is consistent across different viewport sizes.

**Did you introduce any new environment variables? If so, call them out explicitly here:**
No new environment variables were introduced.

ekzyis reviewed (commented):
Welcome to the team! :heart: 

It worked well during some basic testing. I intend to test more tomorrow but I don't expect it to break in any way since it's very similar to how user suggestions work.

However, it seems to also duplicate a lot of code, but I don't expect contributors to refactor our code before they add something—especially not first-time contributors—and I actually like that you didn't since it would have made this a lot harder to review.

However, we might want to have this refactored before we merge this since I assume most of the code could be shared by `UserSuggest` and `TerritorySuggest`.

Would you be interested in doing that as part of this PR?

Else I think merging this would be blocked by someone of us refactoring the foundational code first and then rebasing your changes on that.

I also left a comment asking what the changes in `UserSuggest` are for.

m0wer commented on components/form.js:
```diff
@@ -633,7 +677,12 @@ export function UserSuggest ({
+      setSuggestions(INITIAL_SUGGESTIONS)
+    }
+  }, [isMouseDown])
 
```
I thought it would improve the UX but it's not really needed. In fact it caused a bug: dropdown not closing when selecting an item with a mouse click.

Removed.

ekzyis replied:
How was this meant to improve UX? Like what UX issue did you try to solve?

m0wer replied:
Nevermind. It was to prevent the dropdown from closing prematurely but that does not happen on the main branch or this branch without those changes. So not needed (and is now removed).

m0wer commented:
Thanks :-) Here's the refactor summary:

1. **Created a new `BaseSuggest` component** handling all common functionality:
   - Suggestion state management
   - Keyboard navigation
   - Dropdown UI
   - Item filtering and transformation

2. **Simplified `UserSuggest` and `TerritorySuggest`** by having them pass appropriate parameters to `BaseSuggest`

3. **Added a `BaseInputSuggest` component** that abstracts common functionality from `InputUserSuggest` and `InputTerritorySuggest`

4. **Cleaned up code**:
   - Simplified query handling
   - Removed unnecessary mouse event handlers
   - Made `resetSuggestions` more concise

I've tested accepting autocomplete suggestions with tab, arrows+enter/tab, and mouse click for both usernames and territories. Everything works properly: dropdown appears in the correct position and remains there, autocompletion gets inserted, and dropdown closes afterward.

Let me know if you need anything else!

ekzyis commented:
Can you fix the [lint issues](https://github.com/stackernews/stacker.news/actions/runs/14634042761/job/41103140084?pr=2124)?

m0wer commented:
Lint fixed. Sorry for missing that.

m0wer commented:
Hi @ekzyis ! Anything else to be done here?

ekzyis commented:
Sorry, I got distracted. Will review again today!

huumn commented on components/form.js:
```diff
@@ -211,6 +212,19 @@ export function MarkdownInput ({ label, topLevel, groupClassName, onChange, onKe
     innerRef.current.focus()
   }, [mention, meta?.value, helpers?.setValue])
 
+  const [territory, setTerritory] = useState()
+  const insertTerritory = useCallback((name) => {
+    if (territory?.start === undefined || territory?.end === undefined) return
+    const { start, end } = territory
+    setTerritory(undefined)
+    const first = `${meta?.value.substring(0, start)}~${name}`
+    const second = meta?.value.substring(end)
+    const updatedValue = `${first}${second}`
+    helpers.setValue(updatedValue)
+    setSelectionRange({ start: first.length, end: first.length })
+    innerRef.current.focus()
+  }, [territory, meta?.value, helpers?.setValue])
+
```
This shares a lot of code with the above.

huumn commented on components/form.js:
```diff
@@ -257,12 +272,27 @@ export function MarkdownInput ({ label, topLevel, groupClassName, onChange, onKe
           left: `${left}px`
         }
       })
+      setTerritory(undefined)
+    } else if (/^~\w*$/.test(currentSegment)) {
+      const { top, left } = textAreaCaret(e.target, e.target.selectionStart)
+      setTerritory({
+        query: currentSegment,
+        start: priorSpace + 1,
+        end: nextSpace,
+        style: {
+          position: 'absolute',
+          top: `${top + Number(window.getComputedStyle(e.target).lineHeight.replace('px', ''))}px`,
+          left: `${left}px`
+        }
+      })
+      setMention(undefined)
```
Same. This shares a lot of code with the above.

huumn reviewed (changes requested):
There's still quite a few instances with code duplication that should be pretty straightforward to reduce/share.

m0wer commented:
Refactor done in 1b60cdf. Ready for another round of QA!

huumn commented on components/adv-post-form.js:
```diff
@@ -198,7 +198,7 @@ export default function AdvPostForm ({ children, item, sub, storageKeyPrefix })
         ['nym', 'pct'].forEach(key => {
           const value = window.localStorage.getItem(`${storageKeyPrefix}-forward[${i}].${key}`)
-          if (value) {
+          if (value !== undefined && value !== null) {
```
This is a bug fix unrelated to this PR, but I noticed it while testing so I fixed it.

huumn commented on components/form.js:
```diff
@@ -139,6 +140,94 @@ function setNativeValue (textarea, value) {
   textarea.dispatchEvent(new Event('input', { bubbles: true, value }))
 }
 
+function useEntityAutocomplete ({
```
this is a hook, but it was named like a component `EntityAutocomplete` which made it confusing

huumn commented on components/form.js:
```diff
@@ -617,34 +691,34 @@ function InputInner ({
-      // remove both the leading @ and any @domain after nym
-      const q = query?.replace(/^[@ ]+|[ ]+$/g, '').replace(/@[^\s]*$/, '')
+      // remove the leading character and any trailing spaces
+      const q = query?.replace(/^[@ ~]+|[ ]+$/g, '').replace(/@[^\s]*$/, '').replace(/~[^\s]*$/, '')
```
You had removed this trailing replacement. I'm not sure why it was there, but it's best if we keep it.

huumn commented on components/form.js:
```diff
@@ -737,19 +810,85 @@ export function InputUserSuggest ({
             autoComplete='off'
             onChange={(formik, e) => {
               onChange && onChange(formik, e)
+              if (e.target.value === ovalue) {
+                // we don't need to set the ovalue or query if the value is the same
+                return
+              }
```
This is a fix that I made that's unrelated to this PR. It was a bug where a selecting a suggestion, would re-query because of a race between `onSelect` and `onChange`.

huumn commented on components/form.js:
```diff
@@ -737,19 +810,85 @@ export function InputUserSuggest ({
+  )
+}
+
+export function InputTerritorySuggest ({
+  transformSub, filterSubs, ...props
+}) {
+  return (
+    <BaseInputSuggest
+      transformItem={transformSub}
+      filterItems={filterSubs}
+      SuggestComponent={TerritorySuggest}
+      prefixRegex={/^[~ ]+|[ ]+$/g}
+      {...props}
+    />
+  )
+}
+
+function UserSuggest ({
+  transformUser = user => user, filterUsers = () => true,
+  children, ...props
+}) {
+  return (
+    <BaseSuggest
+      transformItem={transformUser}
+      filterItems={filterUsers}
+      getSuggestionsQuery={USER_SUGGESTIONS}
+      itemsField='userSuggestions'
+      {...props}
+    >
+      {children}
+    </BaseSuggest>
+  )
+}
+
+function TerritorySuggest ({
+  transformSub = sub => sub, filterSubs = () => true,
+  children, ...props
+}) {
+  return (
+    <BaseSuggest
```
The only change I made to these were removing all the parameters that weren't needed due to `...props`.

huumn reviewed (approved):
Nice work! Thanks for sticking with it.

huumn commented:
Do you have a lightning address I can send sats to?

m0wer commented:
Thanks for the improvements and the explanations.

klk@stacker.news happy to be labeled as a contributor in SN as well.

huumn commented:
> happy to be labeled as a contributor in SN as well

Please open a PR adding your nym to the list: https://github.com/stackernews/stacker.news/blob/master/contributors.txt

m0wer commented:
> > happy to be labeled as a contributor in SN as well
> 
> Please open a PR adding your nym to the list: https://github.com/stackernews/stacker.news/blob/master/contributors.txt

https://github.com/stackernews/stacker.news/pull/2158
